waspc/data/Generator/templates/server/src/routes/auth/passport/google/google.js

```diff
@@ -0,0 +1,83 @@
+import express from 'express'
+import passport from 'passport'
+import GoogleStrategy from 'passport-google-oauth20'
+
+import waspServerConfig from '../../../../config.js'
+import { contextWithUserEntity, authConfig } from '../../utils.js'
+import { sign } from '../../../../core/auth.js'
+import { configFn, onSignInFn } from './googleConfig.js'
+
+const config = validateConfig(configFn())
+
+passport.use('waspGoogleStrategy', new GoogleStrategy({
+  clientID: config.clientId,
+  clientSecret: config.clientSecret,
+  callbackURL: `${waspServerConfig.frontendUrl}/auth/redirect/google`,
+  scope: config.scope,
+  passReqToCallback: true
+}, oauthCodeValidationSucceeded))
+
+// This function is invoked after we successfully exchange the one-time-use OAuth code for a real Google API token.
+// This token was used to get the Google profile information supplied as a parameter.
+async function oauthCodeValidationSucceeded(req, _accessToken, _refreshToken, profile, done) {
+  try {
+    const user = await onSignInFn(contextWithUserEntity, { profile })
+
+    if (!user?.id) {
+      return done(new Error('auth.onSignInFn must return a user object with an id property'))
+    }
+
+    // Pass along the userId so we can create the JWT in the OAuth code validation route handler.
+    req.wasp = { ...req.wasp, userId: user.id }
+
+    done(null, user)
+  } catch (err) {
+    return done(err)
+  }
+}
+
+function validateConfig(config) {
+  if (!config?.clientId) {
+    throw new Error("auth.google.configFn must return an object with a clientId property.")
+  }
+
+  if (!config?.clientSecret) {
+    throw new Error("auth.google.configFn must return an object with a clientSecret property.")
+  }
+
+  if (!config?.scope) {
+    throw new Error("auth.google.configFn must return an object with a scope property.")
+  } else if (!Array.isArray(config.scope) || !config.scope.includes('email') || !config.scope.includes('profile')) {
+    throw new Error("auth.google.configFn returned an object with an invalid scope property. It must be an array including 'email' and 'profile'.")
+  }
+
+  return config
+}
+
+const router = express.Router()
+
+// Constructs a Google OAuth URL and redirects browser to start sign in flow.
+router.get('/login', passport.authenticate('waspGoogleStrategy', { session: false }))
+
+// Validates the OAuth code from the frontend, via server-to-server communication
+// with Google. If valid, provides frontend a response containing the JWT.
+// NOTE: `oauthCodeValidationSucceeded` is invoked as part of the `passport.authenticate`
+// call, before the final route handler callback. This is how we gain access to `req.wasp`. 
+router.get('/validateCode',
+  passport.authenticate('waspGoogleStrategy', {
+    session: false,
+    failureRedirect: waspServerConfig.frontendUrl + authConfig.failureRedirectPath
+  }),
+  async function (req, res) {
+    const userId = req.wasp.userId
```

Comment:
How come we don't need optional chaining (i.e., `req.wasp?.userId`)?

Reply:
Good question. This is because `oauthCodeValidationSucceeded` will run before this callback, meaning that we will either have a middleware error, or `req.wasp.userId` will be set. However, it may be ok to add optional chaining since there is an `if` check below, but it feels like if we are confident this should never be the case, that blowing up is probably the best solution here since it indicates some larger bug/issue we don't want to hide.

Reply:
I'm going to respond to both related comments here.

This makes sense, I had a feeling that this could be the reasoning. Btw, and I learned this today by complete accident, including optional chaining here (and input validation in processing code in general) would probably qualify as an antipattern called "Shotgun parsing":

> Shotgun parsing is a programming antipattern whereby parsing and input-validating code is mixed with and spread across processing code—throwing a cloud of checks at the input, and hoping, without any systematic justification, that one or another would catch all the “bad” cases. Shotgun parsing is a programming antipattern whereby parsing and input-validating code is mixed with and spread across processing code—throwing a cloud of checks at the input, and hoping, without any systematic justification, that one or another would catch all the “bad” cases.

Source: https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/#footnote-1

Reply:
Ah, I love that term. I never heard it before, but it makes sense. Thanks for following up and sharing!

In this case though, I think your suggestion would still be valid and fine too. I will probably still do it for at least one case in the `UsernameAndPassword` refactor. 👍🏻

Reply:
Ok, you can resolve the conversation when you decide on the approach fit :smile:

Reply:
I ended up using optional chaining, but followed by error checks/messages on the results 👍🏻